tests/components/google/test_init.py

```diff
@@ -1,67 +1,513 @@
+        yield mock_flow
 
 
-async def test_get_calendar_info(hass, test_calendar):
-    """Test getting the calendar info."""
-    calendar_info = await hass.async_add_executor_job(
-        google.get_calendar_info, hass, test_calendar
+@pytest.fixture
+async def token_scopes() -> list[str]:
+    """Fixture for scopes used during test."""
+    return ["https://www.googleapis.com/auth/calendar"]
+
+
+@pytest.fixture
+async def creds(token_scopes: list[str]) -> OAuth2Credentials:
+    """Fixture that defines creds used in the test."""
+    token_expiry = utcnow() + datetime.timedelta(days=7)
+    return OAuth2Credentials(
+        access_token="ACCESS_TOKEN",
+        client_id="client-id",
+        client_secret="client-secret",
+        refresh_token="REFRESH_TOKEN",
+        token_expiry=token_expiry,
+        token_uri="http://example.com",
+        user_agent="n/a",
+        scopes=token_scopes,
     )
-    assert calendar_info == {
-        "cal_id": "qwertyuiopasdfghjklzxcvbnm@import.calendar.google.com",
-        "entities": [
-            {
-                "device_id": "we_are_we_are_a_test_calendar",
-                "name": "We are, we are, a... Test Calendar",
-                "track": True,
-                "ignore_availability": True,
-            }
-        ],
-    }
-
-
-async def test_found_calendar(hass, google_setup, mock_next_event, test_calendar):
-    """Test when a calendar is found."""
-    config = {
-        "google": {
-            CONF_CLIENT_ID: "id",
-            CONF_CLIENT_SECRET: "secret",
-            "track_new_calendar": True,
+
+
+@pytest.fixture
+async def mock_exchange(creds: OAuth2Credentials) -> YieldFixture[Mock]:
+    """Fixture for mocking out the exchange for credentials."""
+    with patch(
+        "oauth2client.client.OAuth2WebServerFlow.step2_exchange", return_value=creds
+    ) as mock:
+        yield mock
+
+
+@pytest.fixture
+async def token_file(
+    hass: HomeAssistant, creds: OAuth2Credentials, token_filename: Path
+) -> None:
+    """Fixture to populate an existing token file."""
+    storage = Storage(token_filename)
+    storage.put(creds)
+
+
+@pytest.fixture
+async def calendars_config() -> list[dict[str, Any]]:
+    """Fixture for tests to override default calendar configuration."""
+    return [
+        {
+            "cal_id": CALENDAR_ID,
+            "entities": [
+                {
+                    "device_id": "backyard_light",
+                    "name": "Backyard Light",
+                    "search": "#Backyard",
+                    "track": True,
+                }
+            ],
         }
-    }
-    assert await async_setup_component(hass, "google", config)
+    ]
+
+
+@pytest.fixture
+async def calendars_yaml(
+    hass: HomeAssistant,
+    calendars_config: list[dict[str, Any]],
+    yaml_devices_filename: Path,
+) -> None:
+    """Fixture that prepares the calendars.yaml file."""
+    with open(yaml_devices_filename, "w") as out:
```

Comment:
We don't want to write real files I think. We could patch this out with some memory buffer if we really want to test the writing and reading from buffer. Otherwise, just patch the YAML loading/dumping and use a dict instead.

Reply:
Yeah, the existing tests patch this out completely, so the idea is to actually increase test coverage.  So we'd prefer to patch the `open` call in the integration? for example https://github.com/home-assistant/core/blob/a4904bd9bc3bcd80e5f688e4cd017fb25594e397/homeassistant/components/google/__init__.py#L426

Reply:
That could work. If it's tricky I don't think it's too bad if we patch one level higher.

Hopefully these files will go away when we can implement a trigger platform for calendars.

Reply:
Yeah, i realize this is a lot of test code for something we plan to gut. It's just helpful for me to learn this code, and i never manually test anything. My impression is that the google calendar auth code performs pretty flaky, so just don't want to make it much worse.

I'll see, and if its bad we can give up on getting the extra test coverage, or i can move it into a separate file.

Reply:
Sounds good.

Reply:
It seems like it wasn't so bad.  However, i noticed the token files are read/written in multiple ways so i re-wrote the token reading code that checks for existing scopes. It now uses Storage.get just like the other code, and looks in the creds object.   I think I approved the code that added that token check, but wasn't as familiar with the code at the time so this is my fault.